In [1]:
import pandas as pd, math, re, collections
from random import randint


As an exercise, use get to write histogram more concisely. You should be able to eliminate the if statement.

In [2]:
# Original
def histogram(s):
    d = dict()
    for c in s:
        if c not in d:
            d[c] = 1
        else:
            d[c] += 1
    return d


histogram('brontosaurus')

{'b': 1, 'r': 2, 'o': 2, 'n': 1, 't': 1, 's': 2, 'a': 1, 'u': 2}

In [3]:
# Rewritten to eliminate if statement using d.get(key, <what to use instead>)
def xhistogram(s):
    d = dict()
    for c in s:
        d [c] = d.get(c, 0) + 1
    return d

xhistogram('brontosaurus')

{'b': 1, 'r': 2, 'o': 2, 'n': 1, 't': 1, 's': 2, 'a': 1, 'u': 2}

In [4]:
# from collections import defaultdict
d = {}
for c in "abcb":
    d[c] = d.get(c,0) + 1
d

{'a': 1, 'b': 2, 'c': 1}

In [5]:
dict(a="a", b=1)

{'a': 'a', 'b': 1}

In [6]:
# d['f'] = d.get('f', list).append('5')
# TypeError                                 Traceback (most recent call last)
# <ipython-input-16-1e9579882f1c> in <module>
# ----> 1 d['f'] = d.get('f', list).append('5')

# TypeError: descriptor 'append' for 'list' objects doesn't apply to a 'str' object

In [7]:
d

{'a': 1, 'b': 2, 'c': 1}

In [8]:
# del words

## Ex 11.1  
Write a function that reads the words in words.txt and stores them as keys in a dictionary. It doesn’t matter what the values are. Then you can use the in operator as a fast way to check whether a string is in the dictionary.

If you did Exercise 10, you can compare the speed of this implementation with the list in operator and the bisection search.


In [9]:
# Read this once to use throughout the rest of the exercises
with open("words.txt") as fin:
    words = [ w.rstrip() for w in fin.readlines() ]
   

In [10]:
for i, w in enumerate(words):
    print (i, w)
    if i>10:
        break

0 aa
1 aah
2 aahed
3 aahing
4 aahs
5 aal
6 aalii
7 aaliis
8 aals
9 aardvark
10 aardvarks
11 aardwolf


```
trial_target = "vnht"
print("in:         Builtin operator")
%timeit  trial_target in words
print("xin_bisect: Recursive")
%timeit  xin_bisect(trial_target, words)
print("in_bisect:  move pointers to same list")
%timeit  in_bisect(trial_target, words)
print("bisearch:   move pointers to same list, recursive calls")
%timeit  bisearch(trial_target, words, 0, len(trial_target)-1)

in:         Builtin operator
1.05 ms ± 5.57 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
xin_bisect: Recursive
391 µs ± 27.4 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
in_bisect:  move pointers to same list
2.43 µs ± 25.8 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
bisearch:   move pointers to same list, recursive calls
568 ns ± 1.84 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

In [11]:
%timeit "vnht" in words

1.06 ms ± 9.29 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


See that it's time of 22.1 ns is 26x faster than the best bisearch of 568 ns

## Ex 11.2  
Read the documentation of the dictionary method setdefault and use it to write a more concise version of invert_dict. Solution: http://thinkpython2.com/code/invert_dict.py.

In [12]:
# Original version
def invert_dict(d):
    inverse = dict()
    for key in d:
        val = d[key]
        if val not in inverse:
            inverse[val] = [key]
        else:
            inverse[val].append(key)
    return inverse

In [13]:
# New version using setdefault
# Point here is make the value for "key" a list so new value can be appended
def new_invert_dict(d):
    inverse = dict()
    for key in d:
        val = d[key]
        inverse.setdefault(val, []).append(key)
    return inverse

In [14]:
# Use collections
from collections import defaultdict
def defaultdict_invert_dict(d):
    inverse = defaultdict(list)
    for key in d:
        val = d[key]
        inverse[val].append(key)
    return inverse

In [15]:
def test(invert_dict):
    if __name__ == '__main__':
#         print("\n",invert_dict.__name__)
        d = dict(a=1, b=2, c=3, z=1)
        inverse = invert_dict(d)
        for val in inverse:
            keys = inverse[val]
#             print(val, keys)


In [16]:
%timeit test(invert_dict)
%timeit test(new_invert_dict)
%timeit test(defaultdict_invert_dict)


797 ns ± 10.2 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
864 ns ± 2.41 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
1.06 µs ± 19.4 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


**CONCLUSION** Hard way is fastest, but they are all sort of close. defaultdict is 20-25% slower

## Ex 10.3
Memoize the Ackermann function from Exercise 2 and see if memoization makes it possible to evaluate the function with bigger arguments. Hint: no. Solution: http://thinkpython2.com/code/ackermann_memo.py.

In [17]:
# Old Ackermann
# Ex 6.2 Ackerman
def ack(m, n):
    
# Assume m and n are in proper range
#     if ( not isinstance(n, int) \
#         or not isinstance(n, int) ) \
#         or m < 0 \
#         or n < 0:
#         print("Please use whole numbers, m>=0 and n>=0.")
#         return
        
    if m == 0:
        return n+1
    if n == 0:
        return ack(m-1, 1)
    return ack(m-1, ack(m, n-1))

m = 3
n = 4

%timeit ack(m, n)
print(f"ack({m}, {n}) = {ack(m,n)}")


766 µs ± 2.76 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
ack(3, 4) = 125


In [18]:
known = {}

In [19]:
def mack(m, n):
    
    if (m,n) in known:
        return known[(m,n)]
    
    if m == 0:
        result = n+1
        known[m, result] = result
        return result
    
    if n == 0:
        result = ack(m-1, 1)
        known[m, result] = result       
        return result
    
    result = ack(m-1, ack(m, n-1))
    known[m, result] = result       
    return result

m = 3
n = 4
%timeit v = mack(m, n)
print(f"mack({m}, {n}) = {mack(m,n)}")

776 µs ± 10.1 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
mack(3, 4) = 125


In [20]:
# %timeit v = mack(4, 5)
# print(f"mack({m}, {n}) = {v}")

# RecursionError: maximum recursion depth exceeded in comparison

## Ex 10.4  
If you did Exercise 7, you already have a function named has_duplicates that takes a list as a parameter and returns True if there is any object that appears more than once in the list.

Use a dictionary to write a faster, simpler version of has_duplicates. Solution: http://thinkpython2.com/code/has_duplicates.py.

In [21]:
# From Ex 10.7
def has_duplicates(lst):
    if len(lst) < 2:
        return False
    if lst[0] in lst[1:]:
        return True
    return has_duplicates(lst[1:])

In [22]:
def has_duplicates1(lst):
    if len(lst) < 2:
        return False
    if lst[0] in lst[1:]:
        return True
    return has_duplicates1(lst[1:])

In [23]:
def has_duplicates2(lst):
    def hd(lst, s):
        if s == e:
            return False
        if lst[s] == lst[s+1]:
            return True
        return hd(lst, s+1)
    e = len(lst)-1           
    if e < 1:
        return False
    return hd(lst, 0)

In [24]:
def has_duplicates3(lst):
    for i in range(len(lst)):
        if lst[i] in lst[i+1:]:
            return True
    return False

        

In [25]:
def has_duplicates4(lst):
    found = {}
    for v in lst:
        if v in found:
            return True
        else:
            found[v] = 1
    return False

        

In [26]:
def has_duplicates5(lst):
    return len(set(lst)) == len(lst)

In [27]:
import random
random.seed(0)  # Uniform results
mx = 1000
f = 100
lst = [random.randint(0, f*mx) for i in range(mx)]


In [28]:
%timeit -n 10 -r 1 has_duplicates4(lst)

11.6 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 10 loops each)


In [29]:
%timeit -n 10 -r 1 has_duplicates(lst)

1.63 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 10 loops each)


In [30]:
%timeit -n 10 -r 1 has_duplicates5(lst)

18.7 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 10 loops each)


In [31]:
has_duplicates(lst)

True

In [32]:
t = sorted(lst)
for i in range(len(t)-1):
    if t[i] == t[i+1]:
        print(t[i], t[i+1])
    

8697 8697
16295 16295
19840 19840
28675 28675
61720 61720
74379 74379
91661 91661
99876 99876


## Ex 11.5  
Two words are “rotate pairs” if you can rotate one of them and get the other (see rotate_word in Exercise 5).

Write a program that reads a wordlist and finds all the rotate pairs. Solution:

In [33]:
def rotate(ch, r=0):
    if not ch.isalpha():
        return ch
    
    if ch.islower():
        return chr((ord(ch)-ord("a")+r)%26+ord("a"))
    else:
        return chr((ord(ch)-ord("A")+r)%26+ord("A"))
    
def rotate_word(str, r=0):
    result = ""
    for c in str:
        result += rotate(c, r)
    return result

def look_rotated(mx=20):

    rotated = {}
        
    c = 0

    for i, word in enumerate(words):
        
        for n in range(1, 26):
            trial = rotate_word(word, r=n)
            if trial in words:
                rotated.setdefault(word, []).append((n, trial))

        c += 1
        if c > mx:
            break
    return rotated


    

In [34]:
look_rotated(mx=1000)

{'aah': [(4, 'eel')],
 'aba': [(19, 'tut')],
 'abet': [(7, 'hila')],
 'abjurer': [(13, 'nowhere')],
 'abo': [(4, 'efs'), (13, 'nob')],
 'aby': [(3, 'deb'), (25, 'zax')],
 'ache': [(6, 'gink')],
 'act': [(24, 'yar')],
 'acts': [(18, 'sulk')]}

## Ex 11.6  
Here’s another Puzzler from Car Talk (http://www.cartalk.com/content/puzzlers):

This was sent in by a fellow named Dan O’Leary. He came upon a common one-syllable, five-letter word recently that has the following unique property. When you remove the first letter, the remaining letters form a homophone of the original word, that is a word that sounds exactly the same. Replace the first letter, that is, put it back and remove the second letter and the result is yet another homophone of the original word. And the question is, what’s the word?
Now I’m going to give you an example that doesn’t work. Let’s look at the five-letter word, ‘wrack.’ W-R-A-C-K, you know like to ‘wrack with pain.’ If I remove the first letter, I am left with a four-letter word, ’R-A-C-K.’ As in, ‘Holy cow, did you see the rack on that buck! It must have been a nine-pointer!’ It’s a perfect homophone. If you put the ‘w’ back, and remove the ‘r,’ instead, you’re left with the word, ‘wack,’ which is a real word, it’s just not a homophone of the other two words.

But there is, however, at least one word that Dan and we know of, which will yield two homophones if you remove either of the first two letters to make two, new four-letter words. The question is, what’s the word?


You can use the dictionary from Exercise 1 to check whether a string is in the word list.

To check whether two words are homophones, you can use the CMU Pronouncing Dictionary. You can download it from http://www.speech.cs.cmu.edu/cgi-bin/cmudict or from http://thinkpython2.com/code/c06d and you can also download http://thinkpython2.com/code/pronounce.py, which provides a function named read_dictionary that reads the pronouncing dictionary and returns a Python dictionary that maps from each word to a string that describes its primary pronunciation.

Write a program that lists all the words that solve the Puzzler. Solution: http://thinkpython2.com/code/homophone.py.

In [37]:
import importlib.util
spec = importlib.util.spec_from_file_location("pronounce.py", \
         "../code/pronounce.py")
#          "/home/bkrawchuk/notebooks/ThinkPython2/code/pronounce.py")
pronounce = importlib.util.module_from_spec(spec)
spec.loader.exec_module(pronounce)



In [38]:
mx = 200000

soundslike = pronounce.read_dictionary()

def get_3words(word):
    return [ word, word[1:], word[0] + word[2:] ]

def get_sounds(word, length=5):
    if len(word) != length:
        return
    w = get_3words(word)
    if w[0] in soundslike:
        if w[1] in soundslike:
            if w[2] in soundslike:
                return [soundslike[x] for x in w]
    return None


for i, word in enumerate(words):
    if i > mx:
        break
    s = get_sounds(word, length=5)
    if s:
        if s[0] == s[1] == s[2]:
            w = get_3words(word)
            print(f"{w[0]}:({s[0]})", f"{w[1]}:({s[1]})", f"{w[2]}:({s[2]})")
print(i)

eerie:(IH1 R IY0) erie:(IH1 R IY0) erie:(IH1 R IY0)
llama:(L AA1 M AH0) lama:(L AA1 M AH0) lama:(L AA1 M AH0)
llano:(L AA1 N OW0) lano:(L AA1 N OW0) lano:(L AA1 N OW0)
scent:(S EH1 N T) cent:(S EH1 N T) sent:(S EH1 N T)
113782
